aptos-move/framework/supra-framework/sources/automation_registry.move

```diff
@@ -0,0 +1,336 @@
+        id: u64,
+        /// The address of the task owner.
+        owner: address,
+        /// The function signature associated with the registry entry.
+        payload_tx: vector<u8>,
+        /// Expiry of the task, represented in a timestamp in second.
+        expiry_time: u64,
+        /// The transaction hash of the request transaction.
+        tx_hash: vector<u8>,
+        /// Max gas amount of automation task
+        max_gas_amount: u64,
+        /// Maximum gas price cap for the task
+        gas_price_cap: u64,
+        /// Registration epoch number
+        registration_epoch: u64,
+        /// Registration epoch time
+        registration_time: u64,
+        /// Flag indicating whether the task is active.
+        is_active: bool
+    }
+
+    #[event]
+    /// Withdraw user's registration fee event
+    struct FeeWithdrawn has drop, store {
+        to: address,
+        amount: u64
+    }
+
+    #[event]
+    /// Update automation gas limit event
+    struct UpdateAutomationGasLimit has drop, store {
+        automation_gas_limit: u64
+    }
+
+    #[event]
+    /// Update duration upper limit event
+    struct UpdateDurationUpperLimit has drop, store {
+        duration_upper_limit: u64
+    }
+
+    #[event]
+    /// Remove automation task registry event
+    struct RemoveAutomationTask has drop, store {
+        id: u64
+    }
+
+    // todo : this function should call during initialzation, but since we already done genesis in that case who can access the function
+    public(friend) fun initialize(supra_framework: &signer) {
+        system_addresses::assert_supra_framework(supra_framework);
+
+        let (registry_fee_resource_signer, registry_fee_address_signer_cap) = account::create_resource_account(
+            supra_framework,
+            REGISTRY_RESOURCE_SEED
+        );
+
+        move_to(supra_framework, AutomationRegistry {
+            current_index: 0,
+            automation_gas_limit: DEFAULT_AUTOMATION_GAS_LIMIT,
+            duration_upper_limit: DEFAULT_DURATION_UPPER_LIMIT,
+            gas_committed_for_next_epoch: 0,
+            registry_fee_address: signer::address_of(&registry_fee_resource_signer),
+            registry_fee_address_signer_cap,
+            tasks: enumerable_map::new_map(),
+        })
+    }
+
+    public(friend) fun on_new_epoch(supra_framework: &signer) acquires AutomationRegistry {
+        system_addresses::assert_supra_framework(supra_framework);
+
+        let automation_registry = borrow_global_mut<AutomationRegistry>(@supra_framework);
+        let ids = enumerable_map::get_map_list(&automation_registry.tasks);
+
+        let current_time = timestamp::now_seconds();
+        let expired_task_gas = 0;
+
+        // Perform clean up and updation of state
+        vector::for_each(ids, |id| {
+            let task = enumerable_map::get_value_mut(&mut automation_registry.tasks, id);
+            if (task.expiry_time < current_time) {
```

Comment:
Here we should take into account also the tasks that are active during  this new epoch but will be already expired for the next epoch. so we need to check `(task.expire_time < current_time + epoch_timeframe)`

Reply:
I assume that if a task does not have sufficient time duration to run within the epoch, and it expires in the middle of the epoch, such a task will not be considered for execution. Is that correct?

Reply:
yes, true it will not be considered expired during that epoch.